### ISSUE TITLE
[rush] rush `deploy` produces incorrect links with `--target` parameter on Mac OS

### ISSUE THREAD

**Issue by dfee** [microsoft/rushstack#1910]
**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

It seems we're generating abs-path symlinks, though I can't pin down where this error is occurring. Here's a log:

```sh
$ rush deploy -s service-a -t /tmp/service-a
# <expected output> 
$ readlink /tmp/service-a/services/service-a/node_modules/koa
../../../../../../tmp/service-a/common/temp/node_modules/.pnpm/registry.npmjs.org/koa/2.11.0/node_modules/koa
# expected: ../../../common/temp/node_modules/.pnpm/registry.npmjs.org/koa
```

**Comment by octogonz** [comment]
Do you have a repro that you can share? Or is it a private project?

**Comment by dfee** [comment]
I do have a repo I can share: https://github.com/dfee/rush-issues/ (`#master` is fine.)

It seems that this error only occurs when specifying `-t`.

**Comment by octogonz** [comment]
Thanks, I'll take a look.

**Comment by octogonz** [comment]
@dfee I am not able to repro this issue. I tried these steps on a Linux box:

```sh
$ git clone https://github.com/dfee/rush-issues/
$ cd rush-issues
$ rush install
$ rush rebuild
$ rush init-deploy -s test -p service-a
$ mkdir /tmp/service-a
$ rush deploy -s test -t /tmp/service-a
```
When I did this:
```
readlink /tmp/service-a/services/service-a/node_modules/koa
```
I got this:
```
 ../../../common/temp/node_modules/.pnpm/registry.npmjs.org/koa
```
I did not get the `../../../../../../` path.  Any idea what I might be doing wrong?

**Comment by dfee** [comment]
Weird... here's the Dockerfile (effectively sh script) I tried to re-produce with, but got the same result as you.

```Dockerfile
FROM node:lts

RUN mkdir /repo
WORKDIR /repo

RUN git clone https://github.com/dfee/rush-issues /repo
ENV IRR="node /repo/common/scripts/install-run-rush.js"

RUN $IRR install
RUN $IRR build -t service-a
RUN mkdir /tmp/service-a && $IRR deploy -s service-a -t /tmp/service-a
RUN echo $(readlink /tmp/service-a/services/service-a/node_modules/koa)
```

Assumably you don't have a Mac you can repro on?

**Comment by octogonz** [comment]
I do have a MacBook.  [big cloud of dust as I blow off the keyboard] heheh

**Comment by octogonz** [comment]
It repros on Mac. I will debug it. Thanks very much @dfee for reporting this!

**Comment by octogonz** [comment]
The problem is caused by `getRealPath()` in [_deploySymlink()](https://github.com/microsoft/rushstack/blob/26a086e1037af7f9713e739f9833b43eb9d70103/apps/rush-lib/src/logic/deploy/DeployManager.ts#L411):

```ts
    const linkInfo: ILinkInfo = {
      kind: originalLinkInfo.kind,
      linkPath: this._remapPathForDeployFolder(originalLinkInfo.linkPath, subdemploymentState),
      targetPath: this._remapPathForDeployFolder(originalLinkInfo.targetPath, subdemploymentState),
    };

    // Has the link target been created yet?  If not, we should try again later
    if (!FileSystem.exists(linkInfo.targetPath)) {
      return false;
    }

    const newLinkFolder: string = path.dirname(linkInfo.linkPath);
    FileSystem.ensureFolder(newLinkFolder);

    // Link to the relative path for symlinks
    const relativeTargetPath: string = path.relative(FileSystem.getRealPath(newLinkFolder), linkInfo.targetPath);
                                                                ^^^^^^^^^^^
```

The `linkInfo` object looks like this:
```
{
  kind: "folderLink", 
  linkPath: "/tmp/deploy/services/service-a/node_modules/koa", 
  targetPath: "/tmp/deploy/common/temp/node_modules/.pnpm/registry.npmjs.org/koa/2.11.0/node_modules/koa"
}
```
And when we call `getRealPath(newLinkFolder)` on Mac OS `/tmp` becomes `/private/tmp`:
```
newLinkFolder 
--> "/tmp/deploy/services/service-a/node_modules"

node_core_library_1.FileSystem.getRealPath(newLinkFolder)
--> "/private/tmp/deploy/services/service-a/node_modules"
```

**Comment by octogonz** [comment]
Seems like we simply need to call `getRealPath()` on both or neither of the `path.relative()` arguments. I'm out of time for right now, but if someone else wants to analyze it more that would help get this fixed faster.

**Comment by octogonz** [comment]
@dfee this is fixed in Rush 5.24.3
